File: Tests/src/OftTestHarness.cs
namespace BlueHeighliner.OpenFrameTransport.Tests;

/// <summary>
/// A connected client/server pair, both established against each other, for use in tests.
/// </summary>
internal sealed class OftPair : IAsyncDisposable
{
    public required IOftListener Listener { get; init; }

    public required IOftConnection ServerConnection { get; init; }

    public required IOftConnection ClientConnection { get; init; }

    public async ValueTask DisposeAsync()
    {
        await this.ClientConnection.DisposeAsync();
        await this.ServerConnection.DisposeAsync();
        await this.Listener.DisposeAsync();
    }
}

/// <summary>
/// Wraps an <see cref="IOftListener"/>, additionally tracking every connection it has accepted that
/// hasn't disconnected yet — since <see cref="IOftListener"/> itself doesn't (see its own doc
/// comment), tests that need to observe accepted-connection counts do it this way instead.
/// </summary>
internal sealed class TrackedListener : IAsyncDisposable
{
    private readonly IOftListener listener;
    private readonly List<IOftConnection> connections = [];
    private readonly object connectionsLock = new();

    private TrackedListener(IOftListener listener)
    {
        this.listener = listener;
        listener.ConnectedHandler = this.OnConnected;
    }

    public static async Task<TrackedListener> Start(IPEndPoint listenEndPoint, OftHostOptions? options = null, CancellationToken cancellationToken = default) =>
        new(await new OftHoster().Host(listenEndPoint, options, cancellationToken).ConfigureAwait(false));

    public IPEndPoint LocalEndPoint => this.listener.LocalEndPoint;

    public IReadOnlyCollection<IOftConnection> Connections
    {
        get
        {
            lock (this.connectionsLock)
            {
                return this.connections.ToArray();
            }
        }
    }

    /// <summary>
    /// Called whenever this listener accepts a new connection, in addition to this type's own
    /// tracking of <see cref="Connections"/>.
    /// </summary>
    public Action<IOftConnection>? OnConnectedExtra { get; set; }

    /// <summary>
    /// Called whenever a connection this listener accepted disconnects, in addition to this type's
    /// own tracking of <see cref="Connections"/> - since each such connection's
    /// <see cref="IOftConnection.DisconnectedHandler"/> is already used internally for that tracking
    /// (and is single-slot), a test that needs its own per-connection disconnected notification goes
    /// through this property instead of assigning the connection's
    /// <see cref="IOftConnection.DisconnectedHandler"/> directly.
    /// </summary>
    public Action<IOftConnection, Exception?>? OnConnectionDisconnectedExtra { get; set; }

    private void OnConnected(IOftConnection connection)
    {
        lock (this.connectionsLock)
        {
            this.connections.Add(connection);
        }

        connection.DisconnectedHandler = exception =>
        {
            lock (this.connectionsLock)
            {
                this.connections.Remove(connection);
            }

            this.OnConnectionDisconnectedExtra?.Invoke(connection, exception);
        };

        this.OnConnectedExtra?.Invoke(connection);
    }

    public async ValueTask DisposeAsync() => await this.listener.DisposeAsync();
}

/// <summary>
/// Shared setup for tests that need a live, established OFT connection over real TCP/TLS on the
/// loopback interface.
/// </summary>
internal static class OftTestHarness
{
    public static readonly TimeSpan DefaultTimeout = TimeSpan.FromSeconds(10);

    public static async Task<OftPair> Establish(
        int maxPacketDataSize = 16384,
        TimeSpan? rekeyInterval = null,
        OftSecurityMode securityMode = OftSecurityMode.ServerAuthentication,
        TimeSpan? pollInterval = null,
        TimeSpan? pollTimeout = null)
    {
        bool needsServerCertificate = securityMode is OftSecurityMode.ServerAuthentication or OftSecurityMode.DualAuthentication;

        OftHostOptions hostOptions = new()
        {
            Info = "server",
            ServerCertificate = needsServerCertificate ? TestCertificate.Create() : null,
            ClientCertificateValidation = (_, _, _, _) => true,
            MaxPacketDataSize = maxPacketDataSize,
            RekeyInterval = rekeyInterval,
            SecurityMode = securityMode,
            PollInterval = pollInterval ?? TimeSpan.FromSeconds(1),
            PollTimeout = pollTimeout ?? TimeSpan.FromSeconds(5),
        };

        IOftHoster hoster = new OftHoster();
        IOftListener listener = await hoster.Host(new IPEndPoint(IPAddress.Loopback, 0), hostOptions);

        TaskCompletionSource<IOftConnection> serverConnectionSource = new(TaskCreationOptions.RunContinuationsAsynchronously);
        listener.ConnectedHandler = connection => serverConnectionSource.TrySetResult(connection);

        OftConnectOptions connectOptions = new()
        {
            Info = "client",
            MaxPacketDataSize = maxPacketDataSize,
            RekeyInterval = rekeyInterval,
            SecurityMode = securityMode,
            PollInterval = pollInterval ?? TimeSpan.FromSeconds(1),
            PollTimeout = pollTimeout ?? TimeSpan.FromSeconds(5),
            ClientCertificates = securityMode == OftSecurityMode.DualAuthentication ? [TestCertificate.Create()] : null,
            ServerCertificateValidation = (_, _, _, _) => true,
        };

        IOftConnector connector = new OftConnector();

        IOftConnection clientConnection = await connector.Connect("127.0.0.1", listener.LocalEndPoint.Port, connectOptions).WaitAsync(DefaultTimeout);
        IOftConnection serverConnection = await serverConnectionSource.Task.WaitAsync(DefaultTimeout);

        return new OftPair
        {
            Listener = listener,
            ServerConnection = serverConnection,
            ClientConnection = clientConnection,
        };
    }

    public static async Task WaitUntil(Func<bool> condition, TimeSpan timeout)
    {
        DateTime deadline = DateTime.UtcNow + timeout;
        while (!condition())
        {
            if (DateTime.UtcNow > deadline)
            {
                throw new TimeoutException("Condition was not met within the timeout.");
            }

            await Task.Delay(20);
        }
    }

    /// <summary>
    /// Binds and immediately releases a loopback port, so a test can dial a port with high
    /// confidence nothing is listening on it.
    /// </summary>
    public static int ReserveFreePort()
    {
        TcpListener probe = new(IPAddress.Loopback, 0);
        probe.Start();
        try
        {
            return ((IPEndPoint)probe.LocalEndpoint).Port;
        }
        finally
        {
            probe.Stop();
        }
    }

    /// <summary>
    /// Dials and TLS-handshakes a raw connection to a local OFT server, without performing the hail
    /// exchange, so a test can drive the OFT wire protocol directly and exercise the server's
    /// handling of malformed or hostile input.
    /// </summary>
    public static async Task<(TcpClient TcpClient, SslStream SslStream, OftFrameStream FrameStream)> RawConnect(int port, CancellationToken cancellationToken = default)
    {
        TcpClient tcpClient = new();
        await tcpClient.ConnectAsync("127.0.0.1", port, cancellationToken).ConfigureAwait(false);
        SslStream sslStream = new(tcpClient.GetStream(), leaveInnerStreamOpen: false, (_, _, _, _) => true);
        await sslStream.AuthenticateAsClientAsync(new SslClientAuthenticationOptions { TargetHost = "127.0.0.1" }, cancellationToken).ConfigureAwait(false);
        return (tcpClient, sslStream, new OftFrameStream(sslStream));
    }

    /// <summary>
    /// Dials a raw, TLS-less connection to a local <see cref="OftSecurityMode.Trusted"/> OFT server,
    /// so a test can drive the OFT wire protocol directly over plain TCP.
    /// </summary>
    public static async Task<(TcpClient TcpClient, OftFrameStream FrameStream)> RawConnectTrusted(int port, CancellationToken cancellationToken = default)
    {
        TcpClient tcpClient = new();
        await tcpClient.ConnectAsync("127.0.0.1", port, cancellationToken).ConfigureAwait(false);
        return (tcpClient, new OftFrameStream(tcpClient.GetStream()));
    }
}

File: Ports/Java/src/test/java/org/blueheighliner/openframetransport/OftTestHarness.java
package org.blueheighliner.openframetransport;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.time.Duration;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

/** Shared setup for tests that need a live, established OFT connection over real TCP/TLS on loopback. */
final class OftTestHarness {
    static final Duration DEFAULT_TIMEOUT = Duration.ofSeconds(10);

    private OftTestHarness() {
    }

    /**
     * Blocks for {@code future} to complete, unwrapping {@link ExecutionException} to rethrow its
     * cause directly (as an {@link Exception}, matching every test method's own {@code throws
     * Exception}) - so an {@code assertThrows(SomeException.class, ...)} written against the old,
     * directly-blocking {@code connect()}/{@code host()} APIs still sees the same exception type
     * now that those return a {@link CompletableFuture} instead.
     */
    static <T> T await(CompletableFuture<T> future) throws Exception {
        try {
            return future.get(DEFAULT_TIMEOUT.toSeconds(), TimeUnit.SECONDS);
        } catch (ExecutionException e) {
            Throwable cause = e.getCause();
            if (cause instanceof Exception exception) {
                throw exception;
            }

            throw e;
        }
    }

    record Pair(OftListener listener, OftConnection serverConnection, OftConnection clientConnection) implements AutoCloseable {
        @Override
        public void close() {
            this.clientConnection.close();
            this.serverConnection.close();
            this.listener.close();
        }
    }

    static Pair establish() throws Exception {
        return establish(16384, null);
    }

    static Pair establish(int maxPacketDataSize, Duration rekeyInterval) throws Exception {
        return establish(maxPacketDataSize, rekeyInterval, OftSecurityMode.SERVER_AUTHENTICATION, Duration.ofSeconds(1), Duration.ofSeconds(5));
    }

    static Pair establish(int maxPacketDataSize, Duration rekeyInterval, OftSecurityMode securityMode, Duration pollInterval, Duration pollTimeout) throws Exception {
        boolean needsServerContext = securityMode == OftSecurityMode.SERVER_AUTHENTICATION || securityMode == OftSecurityMode.DUAL_AUTHENTICATION;

        OftHostOptions hostOptions = OftHostOptions.builder()
                .info("server")
                .sslContext(needsServerContext ? TestCertificates.createServerContext() : null)
                .maxPacketDataSize(maxPacketDataSize)
                .rekeyInterval(rekeyInterval)
                .securityMode(securityMode)
                .pollInterval(pollInterval)
                .pollTimeout(pollTimeout)
                .build();

        OftHoster hoster = OftHoster.create();
        OftListener listener = await(hoster.host(new InetSocketAddress("127.0.0.1", 0), hostOptions));
        CompletableFuture<OftConnection> serverConnectionFuture = new CompletableFuture<>();
        listener.setConnectedHandler(serverConnectionFuture::complete);

        OftConnectOptions connectOptions = OftConnectOptions.builder()
                .info("client")
                .sslContext(needsServerContext ? TestCertificates.createClientContext() : null)
                .maxPacketDataSize(maxPacketDataSize)
                .rekeyInterval(rekeyInterval)
                .securityMode(securityMode)
                .pollInterval(pollInterval)
                .pollTimeout(pollTimeout)
                .build();

        OftConnector connector = OftConnector.create();
        OftConnection clientConnection = await(connector.connect("127.0.0.1", listener.getLocalEndpoint().getPort(), connectOptions));
        OftConnection serverConnection = serverConnectionFuture.get(DEFAULT_TIMEOUT.toSeconds(), TimeUnit.SECONDS);

        return new Pair(listener, serverConnection, clientConnection);
    }

    static void waitUntil(java.util.function.BooleanSupplier condition, Duration timeout) throws TimeoutException {
        long deadline = System.currentTimeMillis() + timeout.toMillis();
        while (!condition.getAsBoolean()) {
            if (System.currentTimeMillis() > deadline) {
                throw new TimeoutException("Condition was not met within the timeout.");
            }

            try {
                Thread.sleep(20);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new RuntimeException(e);
            }
        }
    }

    /** Reserves a TCP port on loopback and immediately releases it, for tests that need a port nothing is listening on. */
    static int reserveFreePort() throws IOException {
        try (ServerSocket socket = new ServerSocket(0)) {
            return socket.getLocalPort();
        }
    }

    /**
     * Wraps an {@link OftListener} and tracks the connections it has accepted, since
     * {@link OftListener} itself doesn't (see its own doc comment) - mirrors the C# reference
     * implementation's test-only {@code TrackedListener}.
     */
    static final class TrackedListener implements AutoCloseable {
        private final OftListener listener;
        private final List<OftConnection> connections = new CopyOnWriteArrayList<>();

        /** Called whenever this listener accepts a new connection, in addition to this type's own tracking of {@link #getConnections}. */
        volatile Consumer<OftConnection> onConnectedExtra;

        /**
         * Called whenever a connection this listener accepted disconnects, in addition to this
         * type's own tracking - since each such connection's disconnected callback is already used
         * internally for that tracking (and is single-slot), a test that needs its own
         * per-connection disconnected notification goes through this field instead of assigning the
         * connection's disconnected callback directly.
         */
        volatile BiConsumer<OftConnection, Throwable> onConnectionDisconnectedExtra;

        private TrackedListener(OftListener listener) {
            this.listener = listener;
            listener.setConnectedHandler(this::onConnected);
        }

        static TrackedListener start(InetSocketAddress listenEndpoint, OftHostOptions options) throws Exception {
            return new TrackedListener(await(OftHoster.create().host(listenEndpoint, options)));
        }

        InetSocketAddress getLocalEndpoint() {
            return this.listener.getLocalEndpoint();
        }

        List<OftConnection> getConnections() {
            return List.copyOf(this.connections);
        }

        private void onConnected(OftConnection connection) {
            this.connections.add(connection);
            connection.setDisconnectedHandler(exception -> {
                this.connections.remove(connection);
                if (this.onConnectionDisconnectedExtra != null) {
                    this.onConnectionDisconnectedExtra.accept(connection, exception);
                }
            });

            if (this.onConnectedExtra != null) {
                this.onConnectedExtra.accept(connection);
            }
        }

        @Override
        public void close() {
            this.listener.close();
        }
    }
}
